### PR TITLE
20211 - alembic script + model updates for amalgamation table additions & changes

### PR DESCRIPTION
**Pull request by tshyun24**
*Issue #:* /bcgov/entity#20211

*Description of changes:*

- add `amalgamating_businesses_version` and `amalgamations_version` tables
image
image

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the lear license (Apache 2.0).

### CODE REVIEW

File: legal-api/migrations/versions/5238dd8fb805_.py [hunk 1]
```diff
@@ -62,11 +62,63 @@ def upgrade():
 
     # enum added after creating table as DuplicateObject error would be thrown otherwise
     op.add_column('amalgamating_businesses', sa.Column('role', role_enum, nullable=False))
-
+    
+    op.create_table(
+        'amalgamations_version',
+        sa.Column('id', sa.Integer(), nullable=False),
```
Comment (argush3) [inline legal-api/migrations/versions/5238dd8fb805_.py:68]: You will need to following the new versioning model.

Take a look at [entity_roles](https://github.com/bcgov/lear/blob/93d6b09e56b7d8cec80e4e158abcd997e33e9f1d/legal-api/migrations/versions/004f7558063f_second.py#L685-L727) as an example.

So a couple of things to make note of:
- versioning tables are suffixed with `_history` as opposed to `_version` now.
- `transaction_id`,  `end_transaction_id` and `operating_type` are no longer used
- keeping the `filing_id` should be sufficient to trace the changes in place of `transaction_id` in the new versioning  approach

File: legal-api/migrations/versions/5238dd8fb805_.py [hunk 2]
```diff
@@ -56,17 +59,75 @@ def upgrade():
...
+        batch_op.create_index(batch_op.f('ix_amalgamating_businesses_filing_id'), ['filing_id'], unique=False)
+    
+    op.create_table(
+        'amalgamations_history',
+        sa.Column('id', sa.Integer(), nullable=False),
+        sa.Column('legal_entity_id', sa.Integer(), nullable=False),
+        sa.Column('filing_id', sa.Integer(), nullable=False),
+        sa.Column('amalgamation_date', sa.TIMESTAMP(timezone=True), nullable=False),
+        sa.Column('court_approval', sa.Boolean(), nullable=False),
+        sa.Column('operation_type', sa.SmallInteger(), nullable=False),
```
Comment (chenhongjing) [inline legal-api/migrations/versions/5238dd8fb805_.py:84]: `operation_type` in history table should be removed too

**Comment by argush3** [comment]
I think there is at least one broken amalgamation test.  Can you take a look when you have a minute?

File: legal-api/migrations/versions/5238dd8fb805_.py [hunk 3]
```diff
@@ -56,17 +59,74 @@ def upgrade():
         sa.Column('foreign_jurisdiction_region', sa.String(length=10), nullable=True),
         sa.Column('foreign_name', sa.String(length=100), nullable=True),
         sa.Column('foreign_identifier', sa.String(length=50), nullable=True),
+        sa.Column('version', sa.Integer(), nullable=False),
+        sa.Column('change_filing_id', sa.Integer(), nullable=False),
```
Comment (chenhongjing) [inline legal-api/migrations/versions/5238dd8fb805_.py:63]: `change_filing_id` is missing in `amalgamations`, `amalgamations_history`, `amalgamating_businesses_history` tables. 👀  It's used in the new versioning approach.

Comment (tshyun24) [reply legal-api/migrations/versions/5238dd8fb805_.py:63]: > `change_filing_id` is missing in `amalgamations`, `amalgamations_history`, `amalgamating_businesses_history` tables. 👀 It's used in the new versioning approach.

For other 3 tables, it has `filing_id`, which is the same as `change_filing_id`. Wondering for `amalgamating_businessws_history`, should I rename to `change_filing_id` or keep `filing_id`?

Comment (argush3) [reply legal-api/migrations/versions/5238dd8fb805_.py:63]: I talked to Vysakh a bit and have a better understanding now. 

This is what we'll need:
* `amalgamations`/`amalgamation_history` tables will need `filing_id` and `change_filing_id`
* `amalgamating_businesses`/`amalgamating_businesses_history` will only need `change_filing_id`

**Review by JazzarKarim — APPROVED** [review]
LGTM

**Review by argush3 — APPROVED** [review]

**Pull request merged**